core/device_plugin_container.cpp.in

```diff
@@ -7,39 +7,57 @@ ${IMPL_INCLUDES_STRING}
 namespace dronecore {
 
 DevicePluginContainer::DevicePluginContainer(DeviceImpl *impl) :
-    _impl(impl),
 ${PLUGIN_CTOR_STRING}
     _plugin_impl_list()
 {
 ${PLUGIN_LIST_APPEND_STRING}
 
-    for (auto it = _plugin_impl_list.begin(); it != _plugin_impl_list.end(); ++it) {
-        (**it).set_parent(_impl);
+    for (auto plugin_impl : _plugin_impl_list) {
+        plugin_impl->set_parent(impl);
     }
 
     init_plugins();
+
+    impl->subscribe_on_discovery(std::bind(&DevicePluginContainer::enable_plugins, this));
+    impl->subscribe_on_timeout(std::bind(&DevicePluginContainer::disable_plugins, this));
 }
 
 DevicePluginContainer::~DevicePluginContainer()
 {
+    disable_plugins();
     deinit_plugins();
 
-${PLUGIN_DTOR_STRING}
-
-    _impl = nullptr;
+    while (!_plugin_impl_list.empty()) {
```

Comment:
I don't want to be too picky, but I still would see this in `erase_plugins()`. The function would be:

```
disable_plugins();
deinit_plugins();
erase_plugins();
```

Which is consistent in terms of abstraction level. Right now, the `while` construct is at a lower abstraction-level as `disable_plugins()` and `deinit_plugins()`.

Reply:
That's correct but then I need the same in the constructor :). I'll give it a try.

Reply:
Check 8d5c277.